modules/debugging/overview.md

```diff
@@ -1,23 +1,229 @@
+
+While it is easier said than done often times approaching it slowly and seeing how data is being mutated is a way 
+to figure our what is the actual problem is the best way to debug your code. 
+
+Still, there are thousands of different reasons that your code could be faulty. 
+
+To help you out here are some tricks to find patterns that help you quickly identify the error. 
+
+
+## Common errors and how to identify them
+
+### Omitted Case
+Description:
+Ommitted case is when you fail to account for one of the cases that data might come in as. 
+
+
+Ways to tell this is the problem:
+1. Data is falling in a catch-all case.
+2. Infinite loops or time out errors.
+3. Data not changing forms resulting in type errors.
+4. Similar data not being altered.
+
+Note:
+Make sure all cases are accounted for and be careful about having a default catch-all
+case for everything.
+
+Note 2: If it blows up in the empty case
+This is often the most overlooked case, but make sure something happens if it's empty
+it will be accounted for. 
+
+Note 3: Incorrectly assuming that there will be no exceptional cases
+There can be anything inputted. Either limit what you are getting inputted OR account 
+for it in your code. Anything that can be accepted as a data point should be accounted for.
+
+
+### Extraneous Assumption
+Description:
+This is when you assume certain things about the input data that may not be true. It 
+usually means you have to change or account for certain things in your data and maybe even
+*clean* your data to make sure it can and will run properly. 
+
+Ways to tell this is the problem:
+1. Resulting data or output is not in the right form.  
+
+Note:
+Never assume that something not specified in the parameters is given to you. Often
+times you will need to make sure everything passed in to you, or any data you receive is
+in the correct form that you expect it to be before working with it.
+
+1. An ordered list
+2. No duplicates of data
+3. Correct or accurate data
+
+### Incorrect order of Input 
+Description:
+The input of the data is not in the order you expected. This could be in many forms
+such as items in a list, order of data in a string, etc.
+
+Ways to tell this is the problem:
+1. Type errors usually inform you of this.
+2. Data has not been changed from original form.
+
+Note:
+Make sure you reorder your input to make sure if fits what you are excpecting before
+you proceed with working with it. 
+
+<!-- ### Incorrect order of Results ISNT THIS SORT OF AN OBVIOUS ERROR THAT DOES'T REALLY REQUIRE CHANGE-->
+
+### Off by one error
+Description:
+These are usually very easy to spot but you are consistantly off by one in your 
+measure or test in something.
+
+Note:
+Generally there is an edge case that is missed or an index that is not being accounted for.
+If there is a patter of always being off by one in your result that means you are probably missing the first
+or last case.
+
+### Incorrect level of Nesting 
```

Comment:
i think this is more generally referring to nested logic rather than just processing a matrix via nested loops.

And there is some piece of logic being at the wrong level of that nesting.

Nesting doesn't just happen in loops.  E.g., it can happen if there are nested conditionals, or across functions, if there are helper functions being called.